pbs/pbsrequest.go

```diff
@@ -130,85 +122,79 @@ func ParsePBSRequest(r *http.Request, cache cache.Cache) (*PBSRequest, error) {
 		pbsReq.TimeoutMillis = uint64(viper.GetInt("default_timeout_ms"))
 	}
 
-	// anything set in cookie overrides what's in bid request? or opposite?
-	pc := ParseUIDCookie(r)
-	if pbsReq.UserIDs == nil {
-		pbsReq.UserIDs = pc.UIDs
-	} else {
-		for bidder, uid := range pc.UIDs {
-			pbsReq.UserIDs[bidder] = uid
-		}
-	}
-
-	// this would be for the shared adnxs.com domain
-	if anid, err := r.Cookie("uuid2"); err == nil {
-		pbsReq.UserIDs["adnxs"] = anid.Value
+	if pbsReq.Device == nil {
+		pbsReq.Device = &openrtb.Device{}
 	}
 
-	pbsReq.Start = time.Now()
-	pbsReq.Bidders = make([]*PBSBidder, 0)
+	// use client-side data for web requests
+	if pbsReq.App == nil {
+		pc := ParseUIDCookie(r)
+		pbsReq.UserIDs = pc.UIDs
 
-	pbsReq.UserAgent = r.Header.Get("User-Agent")
-	pbsReq.Url = r.Header.Get("Referer") // must be specified in the header
+		// this would be for the shared adnxs.com domain
+		if anid, err := r.Cookie("uuid2"); err == nil {
+			pbsReq.UserIDs["adnxs"] = anid.Value
+		}
 
-	// TODO: this should explicitly put us in test mode
-	if r.FormValue("url_override") != "" {
-		pbsReq.Url = r.FormValue("url_override")
-	}
+		pbsReq.Device.UA = r.Header.Get("User-Agent")
 
-	if strings.Index(pbsReq.Url, "http") == -1 {
-		pbsReq.Url = fmt.Sprintf("http://%s", pbsReq.Url)
-	}
+		if r.Header.Get("X-Real-Ip") != "" {
+			pbsReq.Device.IP = r.Header.Get("X-Real-Ip")
+		} else {
+			ip, _, uerr := net.SplitHostPort(r.RemoteAddr)
+			if uerr == nil && net.ParseIP(ip) != nil {
+				pbsReq.Device.IP = ip
+			}
+		}
+		pbsReq.Url = r.Header.Get("Referer") // must be specified in the header
+		// TODO: this should explicitly put us in test mode
+		if r.FormValue("url_override") != "" {
+			pbsReq.Url = r.FormValue("url_override")
+		}
+		if strings.Index(pbsReq.Url, "http") == -1 {
+			pbsReq.Url = fmt.Sprintf("http://%s", pbsReq.Url)
+		}
 
-	url, err := url.Parse(pbsReq.Url)
-	if err != nil {
-		return nil, fmt.Errorf("Invalid URL '%s': %v", pbsReq.Url, err)
-	}
+		url, err := url.Parse(pbsReq.Url)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid URL '%s': %v", pbsReq.Url, err)
+		}
 
-	if url.Host == "" {
-		return nil, fmt.Errorf("Host not found from URL '%v'", url)
-	}
+		if url.Host == "" {
+			return nil, fmt.Errorf("Host not found from URL '%v'", url)
+		}
 
-	pbsReq.Domain, err = publicsuffix.EffectiveTLDPlusOne(url.Host)
-	if err != nil {
-		return nil, fmt.Errorf("Invalid URL '%s': %v", url.Host, err)
-	}
+		pbsReq.Domain, err = publicsuffix.EffectiveTLDPlusOne(url.Host)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid URL '%s': %v", url.Host, err)
+		}
 
-	_, err = cache.GetDomain(pbsReq.Domain)
-	if err != nil {
-		return nil, fmt.Errorf("Invalid URL %s", pbsReq.Domain)
+		_, err = cache.GetDomain(pbsReq.Domain)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid URL %s", pbsReq.Domain)
+		}
+	} else {
+		_, err = cache.GetDomain(pbsReq.App.Bundle)
```

Comment:
This mimics the check for domains needing to be on a whitelist before we'll serve to them. I'm going to comment out for the moment since I'm not sure this check makes sense in the app world.